### PR TITLE
The initial implementation of up-client-zenoh-python

### PR DESCRIPTION
**Pull request by neelam-kushwah**
Implementation based on up-spec 1.5.7

Current Issues:

1. Ttl is not implemented. **Resolved in https://github.com/eclipse-uprotocol/up-client-zenoh-python/pull/3/commits/bddd157aad61116ac40220286a73c1243361441d commit**
2. Attributes version in the attachment is not set. According to the up-specs, the attachment supports non-unique key-value pairs, whereas in Python, dictionaries cannot have duplicate keys. Therefore, to adhere to the specification, we either need an AttachmentBuilder class similar to the one in zenoh-rust, or the keys should be unique. Currently I am attaching only attributes not version.- **Resolved in https://github.com/eclipse-uprotocol/up-client-zenoh-python/pull/3/commits/573bcc837e4a006009d51425b2ad5b433b237821 commit**
3. Payload Format encoding is not implemented. KnownEncoding class is missing in zenoh-python. **Fixed in https://github.com/eclipse-uprotocol/up-client-zenoh-python/pull/3/commits/d223cb65d96061e6ae303f46e272f007d8c7773e commit**
4. Unable to retrieve attachment from query. The attachment parameter is not present in Query in zenoh-python.  **Resolved in** https://github.com/eclipse-zenoh/zenoh-python/pull/201/files#top

### CODE REVIEW

**Comment by evshary** [comment]
@neelam-kushwah I'll check internally with the issues related to zenoh-python

**Comment by evshary** [comment]
BTW, about the Encoding issue, does this work?
https://github.com/eclipse-zenoh/zenoh-python/blob/0.11.0-rc.2/examples/z_put.py#L111

**Comment by neelam-kushwah** [comment]
> BTW, about the Encoding issue, does this work? https://github.com/eclipse-zenoh/zenoh-python/blob/0.11.0-rc.2/examples/z_put.py#L111

 I think this will work. I'll try it and let you know.

**Comment by neelam-kushwah** [comment]
@evshary Is there any nightly build that I can use for zenoh-python?

**Comment by evshary** [comment]
> @evshary Is there any nightly build that I can use for zenoh-python?

Not yet. The fix is merged now. We'll release 0.11.0-rc.3 these days.
You can build from source for test temporarily.

**Comment by evshary** [comment]
OK. Let me update the status in zenoh-python:
For 1 (TTL issue): https://github.com/eclipse-zenoh/zenoh-python/pull/210 should fix it
For 2 and 4 (Attachment issue): https://github.com/eclipse-zenoh/zenoh-python/pull/201 should fix them
For 3 (Encoding issue): https://github.com/eclipse-zenoh/zenoh-python/pull/208 should fix it.

These fixes will be included in Zenoh 0.11.0-rc.3 release. Before the release, you can use the wheels here for the time being.
https://github.com/eclipse-zenoh/zenoh-python/actions/workflows/release.yml

**Review by stevenhartley — APPROVED** [review]
LGTM

File: NOTICE.adoc
Comment (stevenhartley) [inline NOTICE.adoc:1]: @neelam-kushwah please remove this file, it is not needed anymore.

**Review by stevenhartley — CHANGES_REQUESTED** [review]
minor change. @evshary can you please also review the code? Thanks!

File: up_client_zenoh/zenohutils.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    @staticmethod
+    def attachment_to_uattributes(attachment: Attachment) -> UAttributes:
+        try:
+            version = None
+            version_found = False
+            uattributes = None
+
+            items = attachment.items()
+            for pair in items:
+                if pair[0] == b"":
```
Comment (evshary) [inline up_client_zenoh/zenohutils.py]: I think the key in attachment should always be empty. Maybe you don't need the if-condition

File: up_client_zenoh/upclientzenoh.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+from zenoh import open as zenoh_open, Subscriber, Session, Encoding, Config, Query, Sample, Value, Queryable
+
+from up_client_zenoh.zenohutils import ZenohUtils
+
+logger = logging.getLogger(__name__)
+
+
+class UPClientZenoh(UTransport, RpcClient):
+
+    def __init__(self, session: Session, subscriber_map: Dict[Tuple[str, UListener], Subscriber],
```
Comment (evshary) [inline up_client_zenoh/upclientzenoh.py]: Just curious why we need to have these Dicts in the argument.

File: up_client_zenoh/upclientzenoh.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+        # Map the priority to Zenoh
+        priority = ZenohUtils.map_zenoh_priority(attributes.priority)
+        if not priority:
+            msg = "Unable to map to Zenoh priority"
+            print(f"ERROR: {msg}")
+            return UStatus(code=UCode.INVALID_ARGUMENT, message=msg)
+
+        try:
+            # Simulate sending data
+            print(f"Sending data to Zenoh with key: {zenoh_key}")
```
Comment (evshary) [inline up_client_zenoh/upclientzenoh.py]: Debug info? Maybe should not use print here.

File: up_client_zenoh/upclientzenoh.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+            else:
+                msg = "Unable to register callback with Zenoh"
+                print(msg)
+                return UStatus(code=UCode.INTERNAL, message=msg)
+        except Exception as e:
+            msg = "Unable to register callback with Zenoh"
+            print(msg)
+            return UStatus(code=UCode.INTERNAL, message=msg)
+        msg = "Successfully register callback with Zenoh"
+        print(msg)
```
Comment (evshary) [inline up_client_zenoh/upclientzenoh.py]: Debug info

File: up_client_zenoh/upclientzenoh.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+            try:
+                queryable = self.session.declare_queryable(zenoh_key, callback)
+                with self.queryable_lock:
+                    self.queryable_map[(topic.SerializeToString(), listener)] = queryable
+            except Exception as e:
+                msg = "Unable to register callback with Zenoh"
+                print(msg)
+                return UStatus(code=UCode.INTERNAL, message=msg)
+
+        declare_queryable(callback)
```
Comment (evshary) [inline up_client_zenoh/upclientzenoh.py]: Why do we need to create a local function and call it here?
Can't we run what we have in the function directly?

**Review by evshary — COMMENTED** [review]
I would suggest to have test function / example code to test the functionality.
Also, it would be great to update the README.

File: NOTICE.adoc
Comment (neelam-kushwah) [reply NOTICE.adoc:1]: Done

**Comment by neelam-kushwah** [comment]
> I would suggest to have test function / example code to test the functionality. Also, it would be great to update the README.

I will create a separate PR for example scripts. Currently, I have validated it with up-simulator

File: up_client_zenoh/upclientzenoh.py (hunk 2 above)
Comment (neelam-kushwah) [reply up_client_zenoh/upclientzenoh.py]: It's not needed as I am passing an empty dictionary from its new method. I've simplified it now and removed all the dicts from the methods, initializing them at the time of creating the UpclientZenoh object.

File: up_client_zenoh/upclientzenoh.py (hunk 3 above)
Comment (neelam-kushwah) [reply up_client_zenoh/upclientzenoh.py]: Done

File: up_client_zenoh/upclientzenoh.py (hunk 4 above)
Comment (neelam-kushwah) [reply up_client_zenoh/upclientzenoh.py]: Done

File: up_client_zenoh/upclientzenoh.py (hunk 5 above)
Comment (neelam-kushwah) [reply up_client_zenoh/upclientzenoh.py]: Removed since it didn't add any value, as this local function is only used once

**Review by evshary — APPROVED** [review]
LGTM

File: up_client_zenoh/zenohutils.py (hunk 1 above)
Comment (neelam-kushwah) [reply up_client_zenoh/zenohutils.py]: done

**Pull request merged**